merge_dynamic_streams_block.py

```diff
@@ -0,0 +1,109 @@
+from collections import defaultdict
+from nio.block.base import Block
+from nio.util.discovery import discoverable
+from nio.signal.base import Signal
+from nio.properties import VersionProperty, TimeDeltaProperty, BoolProperty, \
+    Property
+from nio.modules.scheduler import Job
+
+from nio.block.mixins.group_by.group_by import GroupBy
+
+
+@discoverable
+class MergeDynamicStreams(GroupBy, Block):
+
+    """ Merge a dynamic number of streams and combine signals together. """
+
+    ttl = TimeDeltaProperty(default={})
+    notify_once = BoolProperty(default=True)
+    stream = Property(name="Stream", default=None, allow_none=True)
+    version = VersionProperty('0.1.0')
+
+    def __init__(self):
+        super().__init__()
+        # Keep track of the last signal for each stream in each group.
+        # Signals are removed from here when ttl expires.
+        # Stream names are never removed.
+        self._last_signal = defaultdict(lambda: defaultdict(lambda: None))
+        # Keep track of the last signal for each stream in each group.
+        # Signals are removed from here when ttl expires and when notified and
+        # notify_once is True. Stream names are never removed.
+        self._signals = defaultdict(lambda: defaultdict(lambda: None))
+        # Store jobs that remove signals after ttl.
+        self._expiration_jobs = defaultdict(lambda: defaultdict(lambda: None))
+
+    def process_group_signals(self, signals, group, input_id):
+        """Return the merged signals to be notified for this group"""
+        output_signals = []
+        # Group signals into streams and then process each one at a time
+        streams = defaultdict(list)
+        for signal in signals:
+            streams[self.stream(signal)].append(signal)
```

Comment:
Doesn't splitting them into streams here take away the order of the signals coming in? Sending `[A, B, A, B]` makes it act more like `[A, A, B, B]` came in, I feel like that's going to change the way signals are notified out of this block.